src/rgw/rgw_data_sync.cc

```diff
@@ -5192,113 +5228,178 @@ int RGWSyncBucketCR::operate(const DoutPrefixProvider *dpp)
                               status_obj, &bucket_status, false, &objv));
       }
     }
     if (retcode < 0) {
+      tn->log(20, SSTR("ERROR: failed to read bucket status object. error: " << retcode));
       return set_cr_error(retcode);
     }
 
     do {
-      tn->log(20, SSTR("sync status for source bucket: " << bucket_status.state));
+      tn->log(20, SSTR("sync status for source bucket: " << bucket_status.state << 
+            ". lease is: " << (bucket_lease_cr ? "taken" : "not taken") << ". stop indications is: " << bucket_stopped));
+
+      if (bucket_status.state == BucketSyncState::Init || 
+          bucket_status.state == BucketSyncState::Stopped ||
+          bucket_stopped) { 
+        // if state is Init or Stopped, we query the remote RGW for ther state
+		    yield call(new RGWReadRemoteBucketIndexLogInfoCR(sc, sync_pair.dest_bucket, &info));
+        if (retcode < 0) {
+          return set_cr_error(retcode);
+        }
+        if (info.syncstopped) {
+          // remote indicates stopped state
+          tn->log(20, "remote bilog indicates that sync was stopped");
+          if (!bucket_lease_cr) {
+            bucket_lease_cr.reset(new RGWContinuousLeaseCR(env->async_rados, env->store, status_obj,
+                                                         lock_name, lock_duration, this));
+            yield spawn(bucket_lease_cr.get(), false);
+            while (!bucket_lease_cr->is_locked()) {
+              if (bucket_lease_cr->is_done()) {
+                tn->log(5, "ERROR: failed to take bucket lease");
+                set_status("lease lock failed, early abort");
+                drain_all();
+                return set_cr_error(bucket_lease_cr->get_ret_status());
+              }
+              tn->log(5, "waiting on bucket lease");
+              yield set_sleeping(true);
+            }
+          }
+          
+          // check if local state is "stopped"
+          yield call(new ReadCR(dpp, env->async_rados, env->svc->sysobj,
+                status_obj, &bucket_status, false, &objv));
+          if (retcode < 0) {
+            tn->log(20, SSTR("ERROR: failed to read status before writing 'stopped'. error: " << retcode));
+            RELEASE_LOCK(bucket_lease_cr);
+            return set_cr_error(retcode);
+          }
+          if (bucket_status.state != BucketSyncState::Stopped) {
+            // make sure that state is changed to stopped localy
+            bucket_status.state = BucketSyncState::Stopped;
+            yield call(new WriteCR(dpp, env->async_rados, env->svc->sysobj,
+                  status_obj, bucket_status, &objv, false));
+            if (retcode < 0) {
+              tn->log(20, SSTR("ERROR: failed to write 'stopped' status. error: " << retcode));
+              RELEASE_LOCK(bucket_lease_cr);
+              return set_cr_error(retcode);
+            }
+          }
+          if (bucket_status.state == BucketSyncState::Incremental) {
```

Comment:
we just set `bucket_status.state = BucketSyncState::Stopped;` in the block above

Reply:
right. i should move this block before the one that update the bucket wide state object